std/traits.d

```diff
@@ -6840,7 +6840,7 @@ enum bool isStaticArray(T) = __traits(isStaticArray, T);
 /**
  * Detect whether type `T` is a dynamic array.
  */
-enum bool isDynamicArray(T) = is(DynamicArrayTypeOf!T) && !isAggregateType!T;
+enum bool isDynamicArray(T) = is(T == U[], U);
```

Comment:
```suggestion
enum bool isDynamicArray(T) = is(T == U[], U) && __traits(getAliasThis, T).length == 0;
```
Unless you mean to change this to allows structs etc.

Reply:
Note the `==`, it already excludes static arrays and structs with `alias this`.

Reply:
Added some tests

Reply:
Ah, I see then.

Reply:
Good. This at least avoids any further successive template insts.

Reply:
This currently passes but fails after this PR:

```d
import std.traits;
enum E : string
{
    a = "a",
    b = "b",
}
static assert(isDynamicArray!E);
```

Reply:
`E` is a unique type which is convertible to `string`, so `isDynamicArray` should return `false`?

Reply:
Aside from being a behavior change it also breaks symmetry with `isStaticArray`:
```d
import std.traits;
enum E : char[1]
{
    a = ['a'],
    b = ['b'],
}
static assert(isStaticArray!E);
```

Reply:
Interesting ... I would consider this a bug in `isStaticArray` given the wording in the spec. But they should indeed be symmetric.

Reply:
Eh, enums are weird. Try this?
```suggestion
template isDynamicArray(T) {
    static if (is(T == U[], U))
    	enum bool isDynamicArray = true;
    else static if (is(T U == enum))
    	enum bool isDynamicArray = isDynamicArray!U;
    else
    	enum bool isDynamicArray = false;
}
```

Reply:
fwiw this is the shortest form I could muster:
```d
enum bool isDynamicArray(T) = is(T == U[], U)
    || { static if (is(T U == enum)) return isDynamicArray!U; else return false; }();
```

Reply:
Thanks for your suggestion, but I think we should determine the "correct" behaviour first.

FWIW, @nordlow's proposed trait yields the same result:

```
import std.traits : isDynamicArray;

enum E : string
{
    a = "a",
    b = "b",
}

pragma(msg, "isStaticArray!E = ", isDynamicArray!E);
pragma(msg, "is(E == T[], T) = ", is(E == T[], T));
pragma(msg, "__traits(isStaticArray, E) = ", __traits(isStaticArray, E));

```

```
.\generated\windows\debug\64\dmd.exe -c -o- .\array.d
isStaticArray!E = true
is(E == T[], T) = false
__traits(isStaticArray, E) = false
```

Reply:
Ping. Still convinced that `isDynamicArray!E` should yield false because `E` is a distinct type (which complies with the spec).

Note that this doesn't affect any project on BuildKite, so iI think this change is fine (with an appropriate changelog entry).